### DISCUSSION TITLE
PDF Portfolio attachments display order

### DISCUSSION THREAD

**Discussion by erand96** [pymupdf/PyMuPDF#1030] [Q&A] [answered]
Good morning!

I am working on a script to manipulate PDF Portfolios. One issue that I'm having relates to the order that the attachments are displayed (Name, date, order.....etc). Is this "sorted by" criteria found anywhere in the PDF metadata? 

Ideally, I would extract the "sorted by" criteria and then use that to reorganize all of the attachments into a singular PDF file in the same order. 

Let me know if I need to provide any additional information!

**Comment by JorjMcKie** [comment]
What is a "portfolio" in PDF?
* an `EmbeddedFiles` object?
* an annotation?
* ???

(Standard) metadata is a ultimately well-defined term: a Python dictionary with a fixed set of keys.

**Reply by erand96** [reply]
Hi Jorj,

Thank you for the quick reply, hope you're having a good Friday.

A Portfolio is essentially one PDF file that contains multiple attachments which can have various formats. Adobe's overview can be found [here](https://helpx.adobe.com/acrobat/using/overview-pdf-portfolios.html).

Below is a screenshot of a portfolio that I currently have open in Adobe.

![image](https://user-images.githubusercontent.com/82478006/115886367-77bc5880-a416-11eb-90f7-e46a7ee56ddc.png)

Additional information on my full project can be found in discussion post #1010

**Reply by JorjMcKie** [reply]
I see. A PDF portfolio is a superset of `EmbeddedFiles` with lots of configurable metadata information.
It used to be supported until some version of MuPDF, **_but no longer is._**
All you can have with PyMuPDF are embedded files, implemented by a set of `Document´ methods starting with names like `embfile_*`. Features include the full set of usual maintenance tasks, but the only metadata per file are what you see in method `embfile_add(...)` - no date, time, author, etc.  or whatever.

**Reply by erand96** [reply]
That makes sense, thank you for the explanation. Any ideas of other utilities that could help me extract that info?

**Reply by JorjMcKie** [reply]
Not at the moment, sorry.
You may be tempted to use PyMuPDF's low-level object access functions `doc.xref_get_keys()` and `doc.xref_get_key()` to dig your own way through the strcuture of a given PDF's portfolio items.
Of course you must know how this structure works on a technical level: which objects may exist, how they point to other objects, etc. The starting point should always be the sub-dictionary / array contained in the PDF's catalog, which is accessible via the xref returned by `doc.pdf_catalog()`. Here is the layout of a PDF catalog showing the 3 " EmbeddedFiles` of this PDF:
```postscript
1 0 obj
<<
  /Type /Catalog
  /Pages 2 0 R
  /Names <<
    /EmbeddedFiles <<
      /Names [ (file1) <<
            /CI <<
            >>
            /EF <<
              /F 3 0 R
            >>
            /F (file1)
            /UF (file1)
            /Desc (file1)
            /Type /Filespec
          >> (file2) <<
            /CI <<
            >>
            /EF <<
              /F 4 0 R
            >>
            /F (file2)
            /UF (file2)
            /Desc (file2)
            /Type /Filespec
          >> (file3) <<
            /CI <<
            >>
            /EF <<
              /F 5 0 R
            >>
            /F (file3)
            /UF (file3)
            /Desc (file3)
            /Type /Filespec
          >> ]
    >>
  >>
  /PageMode /UseAttachments
>>
endobj
```
So the array containing the embedded files has the size 2 * number of embedded files. The first of each item pair is the unique item's name, and the second a PDF dictionary describing the file itself.
You can directly access that array with the said functions like so:
```python
xref = doc.pdf_catalog()
t, array = doc.xref_get_key(xref, ""Names/EmbeddedFiles/Names")
# which puts the following string in variable 'array':
[(file1)<</CI ... >>]
```

**Comment by JorjMcKie** [comment, marked as answer]
I have made a few changes to PyMuPDF's embedded files handling.
Note that the portfolio information for each embedded file is contained in object encoded in xref "collection".
If you look at [this](http://dev.datalogics.com/cookbook/portfolios/portfolio-sample.pdf) example file, the following things will be possible in v1.18.13:
```python
>>> doc=fitz.open("portfolio-sample.pdf")
>>> n=doc.embfile_count()
>>> from pprint import pprint
>>> for i in range(n):
	pprint(doc.embfile_info(i))
	print("-"*10)

	
{'checksum': b'c59313c2a309cb9a235fe2809464c39266c3bc',
 'collection': 50,
 'creationDate': "D:20120928172443-04'00'",
 'desc': '',
 'filename': 'facility_budget.xlsx',
 'length': 9633,
 'modDate': "D:20120928172458-04'00'",
 'name': '<0>facility_budget.xlsx',
 'size': 12019,
 'ufilename': 'facility_budget.xlsx'}
----------
{'checksum': b'6e0c1707112538c2aa65cb9a143f5864e280ba21',
 'collection': 48,
 'creationDate': "D:20120928172410-04'00'",
 'desc': '',
 'filename': 'globalcorp_projectupdate.pptx',
 'length': 772137,
 'modDate': "D:20120928172433-04'00'",
 'name': '<0>globalcorp_projectupdate.pptx',
 'size': 1179462,
 'ufilename': 'globalcorp_projectupdate.pptx'}
----------
{'checksum': b'c384c5a1c3aee2809e74',
 'collection': 198,
 'creationDate': "D:20120928172335-04'00'",
 'desc': '',
 'filename': 'memo.docx',
 'length': 223867,
 'modDate': "D:20120928172359-04'00'",
 'name': '<0>memo.docx',
 'size': 227780,
 'ufilename': 'memo.docx'}
----------
{'checksum': b'33416be2809a5249c2b738c2b0e2809e5b2c767414e28098',
 'collection': 44,
 'creationDate': "D:20120726135901-04'00'",
 'desc': '',
 'filename': 'new_secure_network.pdf',
 'length': 451454,
 'modDate': "D:20120726140215-04'00'",
 'name': '<0>new_secure_network.pdf',
 'size': 535348,
 'ufilename': 'new_secure_network.pdf'}
----------
{'checksum': b'68c2a846c3ad0dc3a1c3a661c38fc2a8c5a1efac81c2b2c39d6ecb9a',
 'collection': 42,
 'creationDate': "D:20120928172712-04'00'",
 'desc': '',
 'filename': 'rendering.jpg',
 'length': 601541,
 'modDate': "D:20120928172824-04'00'",
 'name': '<0>rendering.jpg',
 'size': 614801,
 'ufilename': 'rendering.jpg'}
----------
>>> # look at 'collection' of last file:
>>> print(doc.xref_object(42))
<<
  /adobe:DisplayName (Office Rendering)
  /adobe:Order 5
  /adobe:Summary 43 0 R
>>
>>> # let us see what is in 'adobe:Summary':
>>> print(doc.xref_object(43))
<<
  /D (Architects vision of one of the social areas of the new location.)
  /RichText (<TextFlow whiteSpaceCollapse="preserve" xmlns="http://ns.adobe.com/textLayout/2008"><p><span>Architects vision of one of the social areas of the new location.</span></p></TextFlow>)
>>
>>> 
```

**Reply by erand96** [reply]
Hi Jorj,

Thank you for the in-depth response. That looks very promising! The particular line that caught my eye was 

> `/adobe:Order 5`

I believe this is would help me out greatly in this project. If I can get the "order" field, then I would just need to identify the sorting method used in the document. Different files are sorted by different criteria, so I need to go through each file, identify the sorting method, and then save the attachments in that same order. I'm hoping to dig through the metadata and eventually find a field that displays the "sort by" criteria. 

Below is just a quick example of what I'm talking about with the "sort by" field. Thanks!

![image](https://user-images.githubusercontent.com/82478006/116094490-31ad0200-a66d-11eb-9d85-8846037714b6.png)

**Comment by JorjMcKie** [comment]
There is a pre-version 1.18.13 in the PyMuPDF-Wheels repo. Look in either the Linux or the OSX branch for your config.
This version has the mehods I was using for that little demo.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Reply by erand96** [reply]
Just to follow up on this, I was able to build code around the example provided above and extract the order of the attachments just how I wanted. I associated the extracted order with a letter that was then inserted before the file name. This allowed the merging function to merge all of the attachments alphabetically which is the only way it can merge.

If there is interest, I can post more of the code later this week once it is finished up.

Thanks for the help!

**Reply by JorjMcKie** [reply]
Congratulations!
Sure I am interested. Please feel free to submit a Pull Request in [this](https://github.com/pymupdf/PyMuPDF-Utilities) repo. So it gets published as your contribution.
